pgrx/src/array/flat_array.rs

```diff
@@ -0,0 +1,340 @@
+use crate::datum::{Array, BorrowDatum, Datum};
+use crate::layout::{Align, Layout};
+use crate::memcx::MemCx;
+use crate::nullable::Nullable;
+use crate::palloc::PBox;
+use crate::pgrx_sql_entity_graph::metadata::{
+    ArgumentError, Returns, ReturnsError, SqlMapping, SqlTranslatable,
+};
+use crate::toast::{Toast, Toasty};
+use crate::{layout, pg_sys, varlena};
+use bitvec::ptr::{self as bitptr, BitPtr, BitPtrError, Const, Mut};
+use bitvec::slice::{self as bitslice, BitSlice};
+use core::iter::{ExactSizeIterator, FusedIterator};
+use core::marker::PhantomData;
+use core::ptr::{self, NonNull};
+use core::{ffi, mem, slice};
+
+use super::port;
+use super::{Element, RawArray, Scalar};
+
+/** `pg_sys::ArrayType` and its unsized varlena
+
+# Safety
+`&FlatArray<'_, T>` assumes its tail is the remainder of a Postgres array of element `T`.
+*/
+#[repr(C)]
+#[derive(Debug)]
+pub struct FlatArray<'mcx, T: ?Sized> {
+    scalar: PhantomData<&'mcx T>,
+    head: pg_sys::ArrayType,
+    tail: [u8],
+}
+
+impl<'mcx, T> FlatArray<'mcx, T>
+where
+    T: ?Sized,
+{
+    fn as_raw(&self) -> RawArray {
+        unsafe {
+            let ptr = NonNull::new_unchecked(ptr::from_ref(self).cast_mut());
+            RawArray::from_ptr(ptr.cast())
+        }
+    }
+
+    /// Number of elements in the array, including nulls
+    ///
+    /// Note that for many arrays, this doesn't have a linear relationship with array byte-len.
+    #[doc(alias = "cardinality")]
+    pub fn nelems(&self) -> usize {
+        self.as_raw().len()
+    }
+
+    /// Number of dimensions the array has
+    pub fn ndims(&self) -> usize {
+        self.head.ndim as _
+    }
+
+    pub fn contains_nulls(&self) -> bool {
+        // SAFETY: Constructive validity from ref and function is non-mutating
+        unsafe { pg_sys::array_contains_nulls((&raw const self.head).cast_mut()) }
+    }
+}
+
+#[derive(Debug)]
+pub struct ArrayAllocError {
+    _stuff: (),
+}
```

Comment:
This allows the `match err.kind()` pattern without committing us to a specific internal representation of the error type here, which may be useful if Postgres creates new restrictions on how to make an array.